### ISSUE TITLE
head to head against https://github.com/Swift-AI/NeuralNet-MNIST

### ISSUE THREAD

**Issue by johndpope** [donald-pinckney/SwiftLearn#1]
fyi - I ran into a crash on current project -
did some rejigging to get latest project to build.
https://github.com/johndpope/SwiftLearn

<img width="1384" alt="screen shot 2017-05-14 at 16 02 13" src="https://cloud.githubusercontent.com/assets/289994/26037312/a9f75eb6-38be-11e7-9c54-b24cad777d4a.png">

I got it to build  by changing the thread_num to 1. 

I want to put your code up against https://github.com/Swift-AI/NeuralNet-MNIST to battle it out. 
I think you've done a great job designing wrappers around matrixes.

https://github.com/Swift-AI/NeuralNet-MNIST

        let X_URL = testBundle.url(forResource: "MNIST_X", withExtension: "csv")!
        let Y_URL = testBundle.url(forResource: "MNIST_Y", withExtension: "csv")!
        let hidden_URL = testBundle.url(forResource: "MNIST_layer0_init", withExtension: "csv")!
        let output_URL = testBundle.url(forResource: "MNIST_layer1_init", withExtension: "csv")!

        let hiddenLayer = NeuralNetworkLayer(weights: Matrix(csvURL: hidden_URL),
                                             activationFunction: .sigmoid)
        let outputLayer = NeuralNetworkLayer(weights: Matrix(csvURL: output_URL),
                                             activationFunction: .sigmoid)

It seems that NeuralNet-MNIST has threading out of the box without needing DispatchQueue.concurrentPerform 
<img width="1061" alt="screen shot 2017-05-14 at 16 38 22" src="https://cloud.githubusercontent.com/assets/289994/26037566/b80aaae4-38c3-11e7-857b-db7889066de3.png">

I'm actually wanting to create a wrapper with tensorflow in swift. 
https://github.com/nubbel/swift-tensorflow/issues/10

**Comment by johndpope** [comment]
can use this https://github.com/johndpope/NeuralNet-Handwriting-iOS
as test bed but need to have saved file conform to swift-ai format. 

From what I can test - it doesn't seem that accurate. 

<img width="310" alt="screen shot 2017-05-14 at 17 13 19" src="https://cloud.githubusercontent.com/assets/289994/26037767/9686f030-38c8-11e7-9ab2-d7d0194e7093.png">


also noteworthy.
https://developer.apple.com/library/content/samplecode/MPSCNNHelloWorld/Introduction/Intro.html

**Comment by donald-pinckney** [comment]
Ok, I'll take a look at my threading, and try to fix it from crashing, or remove it completely if I can get it to work out of the box.

As for your digit recognition, you should definitely be able to get much better accuracy, even using the simple 1 hidden layer neural network. If you want to get much higher accuracy, you should use the Convolutional Neural Networks from Apple that you linked: these can get you much higher accuracy than the neural networks that I have or that swift-ai has (which are essentially very similar).
